home/private_dot_local/lib/agent-equipment/agent_equipment/canonical.py

```
@@ -0,0 +1,571 @@
+def _entry_identity(metadata: os.stat_result) -> tuple[int, int, int]:
+    return (
+        metadata.st_dev,
+        metadata.st_ino,
+        stat.S_IFMT(metadata.st_mode),
+    )
+
+
+def _hash_descriptor(file_descriptor: int, role: str, relative_path: str) -> str:
+    before = os.fstat(file_descriptor)
+    if not stat.S_ISREG(before.st_mode):
+        raise ValueError("manifest entries must be regular files")
```

Ruff (0.16.1)

[warning] Avoid specifying long messages outside the exception class

(TRY003)